### PR TITLE
Feature nxp 24251 web ui publishing

### PR DESCRIPTION
**Pull request by guirenard**
T&P went through https://qa.nuxeo.org/jenkins/job/TestAndPush/job/ondemand-testandpush-grenard-maven3/256/ but I went dry run so PR was not created.

### CODE REVIEW

File: nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/resources/OSGI-INF/rendition-operations-contrib.xml [hunk 1]
```diff
@@ -6,6 +6,8 @@
 
     <operation class="org.nuxeo.ecm.platform.rendition.operation.GetRendition" />
     <operation class="org.nuxeo.ecm.platform.rendition.operation.GetContainerRendition" />
+    <operation class="org.nuxeo.ecm.platform.rendition.operation.PublishRendition"  replace="true"/>
```
Comment (efge) [inline nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/resources/OSGI-INF/rendition-operations-contrib.xml]: I'm not sure it's wise to have an operation that replaces another. Why is that needed?

File: nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-publisher/src/main/java/org/nuxeo/ecm/platform/rendition/publisher/RenditionPublicationFactory.java [hunk 2]
```diff
@@ -37,6 +37,7 @@
 import org.nuxeo.ecm.platform.publisher.task.CoreProxyWithWorkflowFactory;
 import org.nuxeo.ecm.platform.rendition.Rendition;
 import org.nuxeo.ecm.platform.rendition.service.RenditionService;
+import org.nuxeo.ecm.platform.rendition.service.RenditionsRemover;
```
Comment (efge) [inline nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-publisher/src/main/java/org/nuxeo/ecm/platform/rendition/publisher/RenditionPublicationFactory.java]: import doesn't seem used

File: nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/io/PublicationJsonEnricher.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            String escapedId = NXQL.escapeString(document.getId());
+            resultCount = wrapper.getSession()
+                                 .queryProjection(
+                                         String.format(org.nuxeo.ecm.platform.rendition.Constants.ALL_PUBLICATION_QUERY,
+                                                 escapedId, escapedId),
+                                         0, 0)
+                                 .size();
+
+        }
+        jg.writeNumberField("resultsCount", resultCount);
```
[the remark this answers is not in the document]
Comment (guirenard) [reply nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/io/PublicationJsonEnricher.java]: Not that I know.

File: nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/io/PublicationJsonEnricher.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+    @Override
+    public void write(JsonGenerator jg, DocumentModel document) throws IOException {
+        jg.writeObjectFieldStart(NAME);
+        int resultCount;
+        try (SessionWrapper wrapper = ctx.getSession(document)) {
+            String escapedId = NXQL.escapeString(document.getId());
+            resultCount = wrapper.getSession()
+                                 .queryProjection(
+                                         String.format(org.nuxeo.ecm.platform.rendition.Constants.ALL_PUBLICATION_QUERY,
```
[the remark this answers is not in the document]
Comment (guirenard) [reply nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/io/PublicationJsonEnricher.java]: :+1:

File: nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-api/src/main/java/org/nuxeo/ecm/platform/rendition/service/RenditionService.java [hunk 5]
```diff
@@ -81,18 +81,34 @@
      */
     Rendition getRendition(DocumentModel doc, String renditionName);
 
+    /**
+     * Same than {@link #getDefaultRendition(DocumentModel, String, boolean, Map)} with store parameter set to false.
+     */
+    /**
+     * Same than {@link #getDefaultRendition(DocumentModel, String, boolean, Map)} with store parameter set to false.
```
[the remark this answers is not in the document]
Comment (guirenard) [reply nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-api/src/main/java/org/nuxeo/ecm/platform/rendition/service/RenditionService.java]: :+1:

File: nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-api/src/main/java/org/nuxeo/ecm/platform/rendition/service/RenditionService.java [hunk 6]
```diff
@@ -81,18 +81,34 @@
      */
     Rendition getRendition(DocumentModel doc, String renditionName);
 
+    /**
+     * Same than {@link #getDefaultRendition(DocumentModel, String, boolean, Map)} with store parameter set to false.
+     */
+    /**
```
[the remark this answers is not in the document]
Comment (guirenard) [reply nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-api/src/main/java/org/nuxeo/ecm/platform/rendition/service/RenditionService.java]: :+1:

File: nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/operation/PublishRendition.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+import org.nuxeo.ecm.platform.rendition.service.RenditionService;
+import org.nuxeo.runtime.api.Framework;
+
+/**
+ * Publish the rendition of a document. If rendition is not given and default rendition option is false, it falls back
+ * on {@link PublishDocument} operation.
+ *
+ * @since 10.3
+ */
+@Operation(id = PublishRendition.ID, category = Constants.CAT_DOCUMENT, label = "Publish Document's Rendition", description = "Publish the input document choosen rendition into the target section. If rendition is not given and default rendition option is false, it falls back on default publishing. Existing proxy is overrided if the override attribute is set. Return the created proxy.", aliases = {
```
[the remark this answers is not in the document]
Comment (guirenard) [reply nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/operation/PublishRendition.java]: :+1:

File: nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/operation/UnpublishAll.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    public void run(DocumentModel doc) {
+        String escapedId = NXQL.escapeString(doc.getId());
+        session.queryProjection(
+                String.format(org.nuxeo.ecm.platform.rendition.Constants.ALL_PUBLICATION_QUERY, escapedId, escapedId),
+                0, 0)
+               .stream()
+               .map(publication -> new IdRef(publication.get(ECM_UUID).toString()))
+               .forEach(session::removeDocument);
+        // XXX in case of published renditions, do we want to clean up the placeless stored rendition?
+        // AFAIK, if a document is published a couple of time with the same rendtion,
```
[the remark this answers is not in the document]
Comment (guirenard) [reply nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/operation/UnpublishAll.java]: :+1:

File: nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/operation/PublishRendition.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+        "Document.PublishRendition" })
+public class PublishRendition {
+
+    public static final String ID = PublishDocument.ID;
```
[the remark this answers is not in the document]
Comment (guirenard) [reply nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/operation/PublishRendition.java]: Do you mean I'll wait for merging this till next week? I prefer to change and merge asap.

File: nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/operation/PublishRendition.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+    @Context
+    protected CoreSession session;
+
+    @Param(name = "renditionName", required = false)
+    protected String renditionName;
+
+    @Param(name = "defaultRendition", required = false)
+    protected boolean defaultRendition = false;
```
[the remark this answers is not in the document]
Comment (guirenard) [reply nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/operation/PublishRendition.java]: :+1:

File: nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/operation/UnpublishAll.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+import org.nuxeo.ecm.automation.core.annotations.Context;
+import org.nuxeo.ecm.automation.core.annotations.Operation;
+import org.nuxeo.ecm.automation.core.annotations.OperationMethod;
+import org.nuxeo.ecm.core.api.CoreSession;
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.api.IdRef;
+import org.nuxeo.ecm.core.query.sql.NXQL;
+
+/**
+ * Unpublish all document's publication.
```
[the remark this answers is not in the document]
Comment (guirenard) [reply nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/operation/UnpublishAll.java]: :+1:

File: nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/operation/UnpublishAll.java [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+ *
+ * @since 10.3
+ */
+@Operation(id = UnpublishAll.ID, category = Constants.CAT_DOCUMENT, label = "Unpublish Document's Publications", description = "Unpublish the input document's publication.")
```
[the remark this answers is not in the document]
Comment (guirenard) [reply nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/operation/UnpublishAll.java]: :+1:

File: nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/service/RenditionServiceImpl.java [hunk 13]
```diff
@@ -535,28 +528,44 @@ private void removeStoredRenditions(Map<String, List<String>> liveDocumentRefsTo
...
     @Override
     public Rendition getDefaultRendition(DocumentModel doc, String reason, Map<String, Serializable> extendedInfos) {
-        if (defaultRenditionDescriptor == null) {
+        return getDefaultRendition(doc, reason, false, extendedInfos);
+    }
+
+    @Override
+    public Rendition getDefaultRendition(DocumentModel doc, String reason, boolean store, Map<String, Serializable> extendedInfos) {
+        DefaultRenditionDescriptor resolvedDesc = null;
+        for (int i = defaultRenditionDescriptors.size() - 1; i > -1; i--) {
```
[the remark this answers is not in the document]
Comment (guirenard) [reply nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/service/RenditionServiceImpl.java]: :+1:

File: nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-publisher/src/main/java/org/nuxeo/ecm/platform/rendition/publisher/RenditionPublicationFactory.java (hunk 2 above)
Comment (guirenard) [reply nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-publisher/src/main/java/org/nuxeo/ecm/platform/rendition/publisher/RenditionPublicationFactory.java]: it actually is lin 97

File: nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/operation/PublishRendition.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (efge) [reply nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/operation/PublishRendition.java]: Merge now but please be ready to revisit this after discussion. I really don't like the fact that an operation would change implementation. It's a symptom that we don't have enough abstraction over the underlying services.

[the remark this answers is not in the document]
Comment (guirenard) [reply nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/operation/PublishRendition.java]: :+1:

File: nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/service/RenditionServiceImpl.java [hunk 14]
```diff
@@ -579,7 +588,31 @@ public Rendition getDefaultRendition(DocumentModel doc, String reason, Map<Strin
...
+                return storedRendition;
+            }
+        }
         return rendition;
     }
 
+    @Override
+    public DocumentModel publishRendition(DocumentModel doc, DocumentModel target, String renditionName,
+            boolean override) {
+        RenditionService rs = Framework.getService(RenditionService.class);
```
Comment (nelsonsilva) [inline nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/service/RenditionServiceImpl.java]: This is the RenditionService

File: nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/service/RenditionServiceImpl.java [hunk 15]
```diff
@@ -579,7 +588,31 @@ public Rendition getDefaultRendition(DocumentModel doc, String reason, Map<Strin
...
+        Rendition rendition = StringUtils.isEmpty(renditionName) ? rs.getDefaultRendition(doc, "publish", true, null)
```
Comment (nelsonsilva) [inline nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/service/RenditionServiceImpl.java]: We should have reasons as constants somewhere

File: nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/service/RenditionServiceImpl.java [hunk 16]
```diff
@@ -579,7 +588,31 @@ public Rendition getDefaultRendition(DocumentModel doc, String reason, Map<Strin
...
+            StoredRendition storedRendition = storeRendition(doc, rendition);
+            if (storedRendition != null) {
+                return storedRendition;
+            }
+        }
         return rendition;
     }
 
+    @Override
+    public DocumentModel publishRendition(DocumentModel doc, DocumentModel target, String renditionName,
```
Comment (nelsonsilva) [inline nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/service/RenditionServiceImpl.java]: This should probably be in the publishing service instead

File: nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/service/RenditionServiceImpl.java [hunk 17]
```diff
@@ -535,28 +528,44 @@ private void removeStoredRenditions(Map<String, List<String>> liveDocumentRefsTo
...
+        DefaultRenditionDescriptor resolvedDesc = null;
+        for (int i = defaultRenditionDescriptors.size() - 1; i >= 0; i--) {
+            DefaultRenditionDescriptor d = defaultRenditionDescriptors.get(i);
+            if ((StringUtils.isEmpty(reason) && StringUtils.isEmpty(d.reason))
+                    || (reason != null && reason.equals(d.reason))) {
+                resolvedDesc = d;
+                break;
+            }
+        }
+        if (resolvedDesc == null) {
```
Comment (nelsonsilva) [inline nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/service/RenditionServiceImpl.java]: Wouldn't it make more sense to retrieve all rendition descriptors with the given reason and evaluate until != null? This would make it easier to set up a custom default rendition for new document types.

File: nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/service/RenditionServiceImpl.java (hunk 16 above)
Comment (guirenard) [reply nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/service/RenditionServiceImpl.java]: Actually not. We wish to be able to simplify publishing service soon, seen with @troger

File: nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/service/RenditionServiceImpl.java (hunk 14 above)
Comment (guirenard) [reply nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/service/RenditionServiceImpl.java]: And?

File: nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/service/RenditionServiceImpl.java (hunk 17 above)
Comment (guirenard) [reply nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/service/RenditionServiceImpl.java]: I am not sure, this will defeat the ability to simply override an existing contribution

Comment (nelsonsilva) [reply nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/service/RenditionServiceImpl.java]: Yes, we'd need to have another key besides reason, but without it I'm not sure how addons will be able to contribute new doc types with dowload or publish renditions...

File: nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/service/RenditionServiceImpl.java (hunk 14 above)
Comment (nelsonsilva) [reply nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/service/RenditionServiceImpl.java]: And you can drop `RenditionService rs = Framework.getService(RenditionService.class);`

Comment (guirenard) [reply nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/service/RenditionServiceImpl.java]: but of course :) tx

File: nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/service/RenditionServiceImpl.java (hunk 17 above)
Comment (guirenard) [reply nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/service/RenditionServiceImpl.java]: you get a point

File: nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/service/RenditionServiceImpl.java [hunk 18]
```diff
@@ -532,77 +533,63 @@ public Rendition getDefaultRendition(DocumentModel doc, String reason, Map<Strin
...
-                    try {
-                        rendition = getRendition(doc, defaultRenditionName);
-                    } catch (NuxeoException e) {
-                        log.error("Unable to use default rendition " + defaultRenditionName, e);
+        ScriptEngine engine;
+        for (int i = defaultRenditionDescriptors.size() - 1; i >= 0; i--) {
+            DefaultRenditionDescriptor desc = defaultRenditionDescriptors.get(i);
+            if ((StringUtils.isEmpty(reason) && StringUtils.isEmpty(desc.reason))
+                    || (reason != null && reason.equals(desc.reason))) {
+                engine = scriptEngineManager.getEngineByName(desc.getScriptLanguage());
```
Comment (nelsonsilva) [inline nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/service/RenditionServiceImpl.java]: nit: can't we reuse the same engine and move this outside the loop? I guess it's safe to assume scripts won't mess up global context

File: nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/test/resources/test-default-rendition-override-contrib.xml [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+
+  <extension target="org.nuxeo.ecm.platform.rendition.service.RenditionService"
+    point="defaultRendition">
+    <defaultRendition reason="publish">
+      <script language="JavaScript">
+        function run() {
+          if (Document.getType() == "Folder") {
+            return 'zipTreeExport';
+          }
+          return null;
```
Comment (nelsonsilva) [inline nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/test/resources/test-default-rendition-override-contrib.xml]: nit: can we omit the return here just to be sure we're handling implicit default return value of `undefined`?

File: nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/service/RenditionServiceImpl.java (hunk 18 above)
Comment (guirenard) [reply nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/service/RenditionServiceImpl.java]: It it true that we just have `javacript` use case for now. But we can potentially have different script language. I'd say no.

**Comment by nelsonsilva** [comment]
👍
